README.md
![Bndr](https://raw.githubusercontent.com/Xpktro/bndr/master/img/logo.png)

This package is basically two things:

* A library for creating glitch art in the form of [Databending](https://en.wikipedia.org/wiki/Databending) via a simple filter chain for the raw input image data. You can manipulate any jpeg/png image file using the builtin filters or construct your own filters.
* A CLI for generating databended images using random a random filter arrangement (and an optional seed value to produce predictable results).


### Requirements

Bndr requires the `Pillow` package (with zlib and libjpeg support). In most cases you would only need to install `libjpeg` and `zlib` system packages since pip installation will deal with Pillow. Please search for instructions on how to install these packages for your OS (there are *plenty* of resources for this across the internet) by yourself.


### Installation

You can install this package from pip simply by running:

    pip install bndr

Or if you clone the source code repository by running:

    python setup.py install


### Bndr CLI

The Bndr CLI is made up of two commands:

* `bndrimg` - Image conversion a random filter chain or and an optional seed.

    **Parameters:**

    * `input` (positional, mandatory): Location of input image. Must be a jpeg or png image.
    * `-o --output`(optional): Location of the output file. If not supplied, will default to `(input filename)_out.png`. Always produces a png image file.
    * `-s --seed` (optional): Seed for generating a specific filter chain. Can be any string or number.
    
    **Usage Examples:**

    Generate a random glitched image given an input called `input.jpg`:

        bndrimg input.jpg

    Generate a random glitched image (named `output.png`) given an input called `input.jpg`:
    
        bndrimg input.jpg -o output.png

    Generate a glitched image given an input called `input.jpg` with a seed value of `value`:
        
        bndrimg input.jpg -s value

* `bndrtxt` - Simple utility to generate random-case and random-symbol-replace texts.

    **Parameters:**

    * `input` (positional, mandatory): Text to be *bent*.
    
    **Usage Examples:**
    
    Bend the string `Hello world.`:

        bndrtxt "Hello world."


#### `bndrimg` Samples
These samples were generated using the same command (on different image files, of course):

    bndrimg input.jpg -s bndr

![Sample 1](https://raw.githubusercontent.com/Xpktro/bndr/master/img/output_sample1.png)

![Sample 2](https://raw.githubusercontent.com/Xpktro/bndr/master/img/output_sample2.png)


### Bndr Library

#### Quickstart

As mentioned in the introduction, Bndr works chaining a series of filters applied to the raw (binary) image data in order to *bend* it. Let's say, for example, that you want to apply the `ADRNBend` filter to an image and save it's results:

```python
from bndr import Bndr, ADRNBend

with open('input.jpg', 'rb') as input_image:
    bndr = Bndr(input_image.read())
    with open('output.jpg', 'wb') as output_image:
        output_data = bndr.process(ADRNBend())
        output_image.write(output_data)
```

This is a simplified version of what the `bndrimg` command does, since some chained filters configurations can severely damage an image and the input image wont be a jpg all times.

Some filters accept configuration values as positional or named arguments, and the `Bndr.process` function can receive an undefined number of filters as arguments to work (it will apply the filters in order of declaration). For example, to apply a 90 degree rotation (via the `Rotate` filter) prior to the `ADRNBend` filter, we would replace the `output_data` assignation with:

```python
from bndr import Bndr, ADRNBend, Rotate

# ...

output_data = bndr.process(Rotate(90), ADRNBend())
```

Currently, Bndr bundles the following list of filters:

* `bndr.ADRNBend(amount=5)`
A databend filter that replaces bytes in the `a-e` range randomly.
    * **`amount`** - Number of times to repeat the replace operation.

* `bndr.ChrisBend(amount=5)`
A databend filter that adds random repeated byte chunks from the same image.
    * **`amount`** - Number of times to repeat the chunk addition operation.

* `bndr.JpglitchBend(amount=randint(0, 99), seed=randint(0, 99), iterations=randint(0, 115))`
A databend filter that replaces random bytes from the image into arbitrary values.
    * **`amount`** - Value to replace into the file.
    * **`seed`** - Value used to fine-tune the positions to be replaced into the file.
    * **`iterations`** - Number of times the replacement operation will be repeated.

* `bndr.ADRNGrayscale()`
An image filter that converts the image into black and white.

* `bndr.Blur()` 
A filter that blurs the input image using a Gaussian blur.

* `bndr.Rotate([amount=choice((0, 90, 180, 270, 360))])` 
A filter that rotates the input image by a given amount which defaults to a random from the group (0, 90, 180, 270, 360)).

* `bndr.Sharpness([factor=random() * 2])`
A filter that enhaces or blurs the edges of the input image by a given factor. A factor of 0.0 gives a blurred image, a factor of 1.0 gives the original image, and a factor of 2.0 gives a sharpened image.

* `bndr.Desaturate([factor=random()])`
A filter that desaturates the input image colors based on a given factor. A factor of 0.0 gives a black and white image. A factor of 1.0 gives the original image.

* `bndr.Min()`
A filter that picks the lowest pixel value in a matrix of 3x3 surrounding every pixel in the image.

* `bndr.Max()`
A filter that picks the largest pixel value in a matrix of 3x3 surrounding every pixel in the image.

* `bndr.Median()`
A filter that picks the median pixel value in a matrix of 3x3 surrounding every pixel in the image.

#### Bndr API

Bndr is designed (which is a pretentious thing to say, since it's *very* simple) to make your writing of databending filters as unobstrusive as possible since it's nanometric-scale api is composed of three classes:


#### `class bndr.Bndr(image)`
Creates an instance of this class for the given raw image data.

* **`image`** - Raw image data. Must be a bytestring, as a result of a binary file read or similar. Images must are expected to be JPEGs. After the creation of the `Bndr` instance, this data will be available as the instance member `self.image`. 

##### `Bndr.process(*filters)`
Applies the filter chain to the given instance.

* **`*filters`** - List of filters to apply, filters must be sub-instances of `BndrFilter`. Returns the raw image data of the filtered image.


#### `class bndr.BndrFilter(*args, **kwargs)`
Creates an instance of this filter for the given arguments and keyword arguments.

* **`*args`** - Positional, optional arguments that will be available upon creation as the instance member `self.args`.
* **` **kwargs `** - Named, optional arguments that will be available upon creation as the instance member `self.kwargs`.

##### `BndrFilter.process(image)`
Handler for image filtering. Must return a bytestring of the (filtered) input image data.

* **`image`** - Raw image data. Must be a bytestring, as a result of a binary file read or similar.


#### `class bndr.PILBndrFilter(*args, **kwargs)`
Creates an instance of this filter for the given arguments and keyword arguments. This is an utility class that wraps the raw image data into a Pillow Image instance.

* **`*args`** - Positional, optional arguments that will be available upon creation as the instance member `self.args`.
* **` **kwargs `** - Named, optional arguments that will be available upon creation as the instance member `self.kwargs`.

##### `PILBndrFilter.process_img(image)`
Handler for image filtering. Must return a Pillow Image instance.

* **`image`** - Image data in the form of a Pillow Image instance.


CHANGELOG
## [1.0.1] 2017-03-20
### Added
- Python 3 support

## [1.0] 2016-04-17
Finished first working version

bndr/bndr.py
# coding:utf-8
import six
from PIL import Image

if six.PY2:
    from cStringIO import StringIO
else:
    from functools import reduce
    from io import BytesIO as StringIO

__all__ = ['Bndr', 'BndrFilter', 'PILBndrFilter']


class Bndr(object):
    def __init__(self, image):
        self.image = image

    def process(self, *filters):
        return reduce(lambda image, f: f(image), filters, self.image)


class BndrFilter(object):
    def __init__(self, *args, **kwargs):
        self.args = args or []
        self.kwargs = kwargs or {}

    def __call__(self, image):
        return self.process(image)

    def process(self, image):
        raise NotImplementedError


class PILBndrFilter(BndrFilter):
    def __call__(self, image):
        self.image = Image.open(StringIO(image))
        self.format = self.image.format
        return super(PILBndrFilter, self).__call__(image)

    def process(self, image):
        self.image = self.process_img(self.image)
        output_str = StringIO()
        self.image.save(output_str,
                        format='JPEG', quality=95)
        output_str.seek(0)
        return output_str.read()

    def process_img(self, image):
        raise NotImplementedError


bndr/bndrimg.py
# coding:utf-8
import argparse
import random
import six

from PIL import Image

if six.PY2:
    from cStringIO import StringIO
else:
    from io import BytesIO as StringIO

from .bndr import Bndr
from .filters import Blur, Sharpness, Desaturate, Min, Max, Median, \
    ADRNGrayscale, Rotate, ADRNBend, ChrisBend, JpglitchBend


def random_filters(seed=None):
    if seed:
        random.seed(seed)
    filters = []

    if random.randint(1, 5) == 1:
        filters.append(Blur())

    if random.randint(1, 7) == 1:
        filters.append(Sharpness())

    if random.randint(1, 4) == 1:
        filters.append(Desaturate())

    if random.randint(1, 10) == 1:
        filters.append(Min())

    if random.randint(1, 10) == 1:
        filters.append(Max())

    if random.randint(1, 10) == 1:
        filters.append(Median())

    if random.randint(1, 30) == 1:
        filters.append(ADRNGrayscale())

    filters = list(filter(lambda i: random.randint(1, 10) == 1, filters))

    for _ in range(random.randint(0, 3)):
        if random.randint(1, 7) <= 2:
            amount = random.choice((0, 45, 90, 135, 180, 225, 270, 315, 360))
            amount = amount * -1 if random.choice((True, False)) else amount
            filters.append(Rotate(amount))
            filters.append(Rotate(amount * -1))

    for _ in range(random.randint(1, 3)):
        filters.append(random.choice([
            ADRNBend(amount=random.randint(3, 20)),
            ChrisBend(amount=random.randint(3, 20)),
            JpglitchBend()
        ]))

    random.shuffle(filters)
    return filters


def output_filename(input_filename):
    filename, extension = input_filename.split('.')
    return '{}_out.png'.format(filename)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('input', help='input image')
    parser.add_argument('-o', '--output', required=False, help='output image')
    parser.add_argument('-s', '--seed', default='', help='seed')
    args = parser.parse_args()

    with open(args.input, 'rb') as input_image:
        jpg_output = StringIO()
        Image.open(input_image).save(jpg_output, format='JPEG', quality=100)
        jpg_output.seek(0)
        bndr = Bndr(jpg_output.read())
        output = args.output or output_filename(args.input)
        with open(output, 'wb') as output_image:
            while True:
                try:
                    filters = random_filters(args.seed)
                    Image.open(StringIO(bndr.process(*filters)))\
                         .save(output_image, format='PNG')
                    break
                except:
                    continue


if __name__ == '__main__':
    main()


bndr/filters.py
# coding:utf-8
import copy
import random
import six

if six.PY2:
    from itertools import izip, tee
    from cStringIO import StringIO
else:
    from io import BytesIO as StringIO
    from itertools import tee
    izip = zip
    xrange = range

from PIL import Image, ImageFilter, ImageEnhance

from .bndr import BndrFilter, PILBndrFilter


__all__ = ['ADRNBend', 'ChrisBend', 'JpglitchBend', 'ADRNGrayscale', 'Blur',
           'Rotate', 'Sharpness', 'Desaturate', 'Min', 'Max', 'Median']


class ADRNBend(BndrFilter):
    """Based on databend.py from adrn https://gist.github.com/adrn/4090186"""

    def process(self, image):
        header = image[:1000]
        core_data = image[1000:]
        data_size = len(core_data)

        letters = 'abcde'
        for _ in range(self.kwargs.get('amount', 5)):
            ii = random.randint(0, data_size - 1)
            jj = random.randint(ii, ii + random.randint(100, 10000))

            pre = core_data[:ii]
            post = core_data[jj:]
            sub_data = core_data[ii:jj]
            sub_data = sub_data.replace(random.choice(letters),
                                        random.choice(letters))

            core_data = pre + sub_data + post
        return header + core_data


class ChrisBend(BndrFilter):
    """Based on an article from Chris Cuellar:
    http://blog.art21.org/2011/09/20/how-to-use-python-to-create-a-simple-flickr-photo-glitcher"""

    def process(self, image):
        image_processed = image[:]
        for i in range(self.kwargs.get('amount', 5)):
            image_processed = self.splice_a_chunk_in_a_file(image_processed)
        return image_processed

    def splice_a_chunk_in_a_file(self, file_data):
        start_point, end_point = \
            self.get_random_start_and_end_points_in_file(file_data)
        section = file_data[start_point:end_point]
        repeated = ''

        for i in range(1, random.randint(1, 5)):
            repeated += section

        new_start_point, new_end_point = \
            self.get_random_start_and_end_points_in_file(file_data)
        file_data = file_data[:new_start_point] + repeated + \
                    file_data[new_end_point:]
        return file_data

    def get_random_start_and_end_points_in_file(self, file_data):
        start_point = random.randint(2500, len(file_data))
        end_point = start_point + \
                    random.randint(0, len(file_data) - start_point)

        return start_point, end_point


class JpglitchBend(BndrFilter):
    """Based on jpglitch.py from https://github.com/Kareeeeem/jpglitch"""
    def process(self, image):
        while True:
            new_bytes = self.glitch_bytes(image)
            try:
                im = Image.open(StringIO(new_bytes))
                im.save(StringIO(), format='JPEG', quality=100)
                return new_bytes
            except IOError:
                pass

    def glitch_bytes(self, image):
        """Glitch the image bytes, after the header based on the parameters.
        'Amount' is the hex value that will be written into the file. 'Seed'
        tweaks the index where the value will be inserted, rather than just a
        simple division by iterations. 'Iterations' should be self explanatory
        """

        amount = self.kwargs.get('amount', random.randint(0, 99)) / 100
        seed = self.kwargs.get('seed', random.randint(0, 99)) / 100
        iterations = self.kwargs.get('iterations', random.randint(0, 115))
        header_length = self.get_header_length(image)

        # work with a copy of the original bytes. We might need the original
        # bytes around if we glitch it so much we break the file.
        new_bytes = bytearray(copy.copy(image))

        for i in (xrange(iterations)):
            max_index = len(image) - header_length - 4

            # The following operations determine where we'll overwrite a value
            # Illustrate by example

            # 36 = (600 / 50) * 3
            px_min = int((max_index / iterations) * i)

            # 48 = (600 / 50) * 3 + 1
            px_max = int((max_index / iterations) * (i + 1))

            # 12 = 48 - 36
            delta = (px_max - px_min)  # * 0.8

            # 36 + (12 * 0.8)
            px_i = int(px_min + (delta * seed))

            # If the index to be changed is beyond bytearray length file set
            # it to the max index
            if px_i > max_index:
                px_i = max_index

            byte_index = header_length + px_i
            new_bytes[byte_index] = int(amount * 256)

        return new_bytes

    def pairwise(self, iterable):
        """Awesome function from the itertools cookbook
        https://docs.python.org/2/library/itertools.html
        s -> (s0,s1), (s1,s2), (s2, s3), ...
        """
        a, b = tee(iterable)
        next(b, None)
        return izip(a, b)

    def get_header_length(self, image):
        """Get the length of the header by searching sequential 0xFF 0xDA
        values. These values mark the end of a Jpeg header. We add two to give
        us a little leeway. We don't want to mess with the header.
        """

        for i, pair in enumerate(self.pairwise(image)):
            if pair[0] == 255 and pair[1] == 218:
                result = i + 2
                return result

        # Fallback from ADRNBend, original raised an exception
        return 1000


class ADRNGrayscale(PILBndrFilter):
    """Based on databend.py from adrn https://gist.github.com/adrn/4090186"""
    def process_img(self, image):
        return image.convert('1')


class Blur(PILBndrFilter):
    def process_img(self, image):
        return image.filter(ImageFilter.GaussianBlur())


class Rotate(PILBndrFilter):
    def process_img(self, image):
        if len(self.args) > 0:
            amount = self.args[0]
        else:
            amount = random.choice((0, 90, 180, 270, 360))
        return image.rotate(amount)


class Sharpness(PILBndrFilter):
    def process_img(self, image):
        if len(self.args) > 0:
            factor = self.args[0]
        else:
            factor = random.random() * 2
        return ImageEnhance.Sharpness(image).enhance(factor)


class Desaturate(PILBndrFilter):
    def process_img(self, image):
        if len(self.args) > 0:
            factor = self.args[0]
        else:
            factor = random.random()
        return ImageEnhance.Color(image).enhance(factor)


class Min(PILBndrFilter):
    def process_img(self, image):
        return image.filter(ImageFilter.MinFilter)


class Max(PILBndrFilter):
    def process_img(self, image):
        return image.filter(ImageFilter.MaxFilter)


class Median(PILBndrFilter):
    def process_img(self, image):
        return image.filter(ImageFilter.MedianFilter)
